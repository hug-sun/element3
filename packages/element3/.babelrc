{
  "presets": [["@babel/preset-env", { "targets": { "node": "current" } }]],
  "plugins": [
    "@vue/babel-plugin-jsx",
    "@babel/plugin-proposal-optional-chaining",
    "@babel/plugin-proposal-nullish-coalescing-operator"
  ],
  "env": {
    "test": {
      "presets": [["@babel/preset-env", { "targets": { "node": "current" } }]]
    },
    "esm": {
      "presets": [["@babel/preset-env", { "modules": false }]]
    }
  }
}
