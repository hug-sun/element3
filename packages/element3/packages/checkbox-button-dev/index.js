import ElCheckboxButton from './CheckboxButton'

ElCheckboxButton.install = (app) => {
  app.component(ElCheckboxButton.name, ElCheckboxButton)
}

export default ElCheckboxButton
