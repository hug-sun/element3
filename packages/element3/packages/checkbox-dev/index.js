import ElCheckbox from './Checkbox'

/* istanbul ignore next */
ElCheckbox.install = function (app) {
  app.component(ElCheckbox.name, ElCheckbox)
}

export default ElCheckbox
