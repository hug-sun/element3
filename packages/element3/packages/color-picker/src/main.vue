<template>
  <div
    :class="[
      'el-color-picker',
      colorDisabled ? 'is-disabled' : '',
      colorSize ? `el-color-picker--${colorSize}` : ''
    ]"
    ref="referenceElm"
    v-clickoutside="hide"
  >
    <div class="el-color-picker__mask" v-if="colorDisabled"></div>
    <div class="el-color-picker__trigger" @click="handleTrigger">
      <span class="el-color-picker__color" :class="{ 'is-alpha': showAlpha }">
        <span
          class="el-color-picker__color-inner"
          :style="{
            backgroundColor: displayedColor
          }"
        ></span>
        <span
          class="el-color-picker__empty el-icon-close"
          v-if="!modelValue && !showPanelColor"
        ></span>
      </span>
      <span
        class="el-color-picker__icon el-icon-arrow-down"
        v-show="modelValue || showPanelColor"
      ></span>
    </div>
    <picker-dropdown
      ref="dropdown"
      :popperClass="popperClass"
      v-model="showPicker"
      @pick="confirmValue"
      @clear="clearValue"
      :color="color"
      :show-alpha="showAlpha"
      :predefine="predefine"
    >
    </picker-dropdown>
  </div>
</template>

<script>
import {
  reactive,
  inject,
  toRefs,
  onMounted,
  computed,
  watch,
  nextTick,
  provide
} from 'vue'

import Color from './color'
import PickerDropdown from './components/picker-dropdown.vue'
import Clickoutside from '../../../src/directives/clickoutside'
import { useEmitter } from '../../../src/composables/emitter'

export default {
  name: 'ElColorPicker',

  props: {
    modelValue: String,
    showAlpha: Boolean,
    colorFormat: String,
    disabled: Boolean,
    size: String,
    popperClass: String,
    predefine: Array
  },
  emits: ['active-change', 'input', 'change', 'update:modelValue'],
  setup(props, context) {
    const color = new Color({
      enableAlpha: props.showAlpha,
      format: props.colorFormat
    })
    const state = reactive({
      color,
      showPicker: false,
      showPanelColor: false,
      popperElm: null,
      referenceElm: null
    })

    provide('referenceState', state)

    inject('elForm', {})
    const { dispatch } = useEmitter()
    const elFormItem = inject('elFormItem', {})
    const elForm = inject('elForm', {})

    const displayedColor = computed(() => {
      if (!props.modelValue && !state.showPanelColor) {
        return 'transparent'
      }
      return displayedRgb(state.color, props.showAlpha)
    })

    const _elFormItemSize = computed(() => {
      return (elFormItem || {}).elFormItemSize
    })

    const colorSize = computed(() => {
      return (
        props.size || _elFormItemSize.value || (context.$ELEMENT || {}).size
      )
    })

    const colorDisabled = computed(() => {
      return props.disabled || (elForm || {}).disabled
    })

    watch(
      () => props.modelValue,
      (val) => {
        if (!val) {
          state.showPanelColor = false
        } else if (val && val !== state.color.value) {
          state.color.fromString(val)
        }
      }
    )

    watch(state.color, () => (state.showPanelColor = true), {
      deep: true
    })

    watch(displayedColor, (val) => {
      if (!state.showPicker) return
      const currentValueColor = new Color({
        enableAlpha: props.showAlpha,
        format: props.colorFormat
      })
      if (val !== currentValueColor) {
        context.emit('active-change', val)
      }
    })
    onMounted(() => {
      const value = props.modelValue
      if (value) {
        state.color.fromString(value)
      }
    })

    function displayedRgb(color, showAlpha) {
      if (!(color instanceof Color)) {
        throw Error('color should be instance of Color Class')
      }

      const { r, g, b } = color.toRgb()
      return showAlpha
        ? `rgba(${r}, ${g}, ${b}, ${color.get('alpha') / 100})`
        : `rgb(${r}, ${g}, ${b})`
    }
    const handleTrigger = () => {
      if (colorDisabled.value) return
      state.showPicker = !state.showPicker
    }

    const confirmValue = () => {
      const value = state.color.value
      context.emit('update:modelValue', value)
      context.emit('change', value)
      dispatch('ElFormItem', 'el.form.change', value)
      state.showPicker = false
    }

    const clearValue = () => {
      context.emit('update:modelValue', null)
      context.emit('change', null)
      if (props.modelValue !== null) {
        dispatch('ElFormItem', 'el.form.change', null)
      }
      state.showPanelColor = false
      state.showPicker = false
      resetColor()
    }

    const hide = () => {
      state.showPicker = false
      resetColor()
    }
    const resetColor = () => {
      nextTick(() => {
        if (props.modelValue) {
          state.color.fromString(props.modelValue)
        } else {
          state.showPanelColor = false
        }
      })
    }
    return {
      ...toRefs(state),
      handleTrigger,
      confirmValue,
      clearValue,
      hide,
      displayedRgb,
      colorSize,
      colorDisabled,
      displayedColor
    }
  },

  directives: { Clickoutside },

  components: {
    PickerDropdown
  }
}
</script>
