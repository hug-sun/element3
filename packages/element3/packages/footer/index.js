import ElFooter from './Footer'

/* istanbul ignore next */
ElFooter.install = function (app) {
  app.component(ElFooter.name, ElFooter)
}

export { ElFooter }

export default ElFooter
