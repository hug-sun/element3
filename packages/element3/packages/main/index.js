import ElMain from './Main'

/* istanbul ignore next */
ElMain.install = function (app) {
  app.component(ElMain.name, ElMain)
}

export { ElMain }
export default ElMain
