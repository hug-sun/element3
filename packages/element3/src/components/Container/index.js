import ElContainer from './src/Container.vue'

/* istanbul ignore next */
ElContainer.install = function (app) {
  app.component(ElContainer.name, ElContainer)
}

export { ElContainer }
