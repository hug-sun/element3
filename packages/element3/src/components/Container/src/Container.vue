<template>
  <section class="el-container" :class="{ 'is-vertical': isVertical }">
    <slot></slot>
  </section>
</template>

<script>
import { computed, toRefs } from 'vue'
export default {
  name: 'ElContainer',
  props: {
    direction: {
      type: String,
      default: 'horizontal'
    }
  },
  setup(props, { slots }) {
    const { direction } = toRefs(props)

    const isVertical = computed(() => {
      if (direction === 'vertical') {
        return true
      } else if (direction === 'horizontal') {
        return false
      }
      if (slots && slots.default) {
        return slots.default().some((vNode) => {
          const tag = vNode.type && vNode.type.name
          return tag === 'ElHeader' || tag === 'ElFooter'
        })
      } else {
        return false
      }
    })
    return {
      isVertical
    }
  }
}
</script>
