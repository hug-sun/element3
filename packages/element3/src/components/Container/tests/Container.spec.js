import Container from '../src/Container.vue'
import { ElHeader } from '../../../../packages/header'
import { ElFooter } from '../../../../packages/footer'
import { ElMain } from '../../../../packages/main'
import { mount } from '@vue/test-utils'

describe('Container.vue', () => {
  it('snapshot', () => {
    const wrapper = mount(Container)
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('direction', () => {
    it('set Container vertical by prop direction', () => {
      const wrapper = mount(Container, {
        props: {
          direction: 'vertical'
        }
      })
      expect(wrapper.find('.is-vertical').exists()).toBe(false)
    })

    it('set Container horizontal by prop direction', () => {
      const wrapper = mount(Container, {
        props: {
          direction: 'horizontal'
        }
      })
      expect(wrapper.find('.is-vertical').exists()).toBe(false)
    })

    it('contain has class el-container', () => {
      const wrapper = mount(Container)
      expect(wrapper.classes('el-container')).toBe(true)
    })
  })

  describe('slot', () => {
    it('slot has el-header ', () => {
      const TestComponent = {
        template: `
        <el-container>
          <el-header></el-header>
          <el-main></el-main>
        </el-container>`,
        components: {
          'el-container': Container,
          'el-header': ElHeader,
          'el-main': ElMain
        }
      }

      const wrapper = mount(TestComponent)
      expect(wrapper.classes('is-vertical')).toBe(true)
    })

    it('slot has  el-footer', () => {
      const TestComponent = {
        template: `
        <el-container>
          <el-footer></el-footer>
          <el-main></el-main>
        </el-container>`,
        components: {
          'el-container': Container,
          'el-footer': ElFooter,
          'el-main': ElMain
        }
      }

      const wrapper = mount(TestComponent)
      expect(wrapper.classes('is-vertical')).toBe(true)
    })
  })
})
