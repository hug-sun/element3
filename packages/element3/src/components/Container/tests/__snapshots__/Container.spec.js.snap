// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container.vue snapshot 1`] = `
<section
  class="el-container"
>
  
  
</section>
`;
